File: 236__Lowest_Common_Ancestor_of_A_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        if root == p or root == q:
            return root
        leftnode, rightnode = None, None
        if root.left:
            # Return some node if the "root" is p or q
            leftnode = self.lowestCommonAncestor(root.left, p, q)
        if root.right:
            # Same as above
            rightnode = self.lowestCommonAncestor(root.right, p, q)

        # Means we find either p or q from one side and another from the other
        # side --> return node
        if leftnode and rightnode:
            return root
        # One of the side find nothing, means that both p and q are in one side,
        # and the ancestor is at the place of the leftnode / rightnode.
        else:
            return leftnode or rightnode

File: Java/236__Lowest_Common_Ancestor_of_a_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    TreeNode result = null;

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        result = null;
        helper(root, p, q);
        return result;
    }

    private boolean helper(TreeNode cur_node, TreeNode p, TreeNode q) {
        if (cur_node == null) return false;
        int left = helper(cur_node.left, p, q) ? 1 : 0;
        int right = helper(cur_node.right, p, q) ? 1 : 0;
        int mid = (cur_node == p || cur_node == q) ? 1 : 0;
        if (left + right + mid >= 2)
            result = cur_node;
        return left + right + mid > 0;
    }
}
